## Homework

In this homework, we'll deploy the bees vs wasps model we trained in the 
[previous homework](../../08-neural-networks-and-deep-learning/homework/homework.md).

Download the model from here: 

https://github.com/alexeygrigorev/large-datasets/releases/download/wasps-bees/bees-wasps.h5


In [2]:
!wget https://github.com/alexeygrigorev/large-datasets/releases/download/wasps-bees/bees-wasps.h5

--2023-11-22 20:37:41--  https://github.com/alexeygrigorev/large-datasets/releases/download/wasps-bees/bees-wasps.h5
Resolving github.com (github.com)... 140.82.121.3
Connecting to github.com (github.com)|140.82.121.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/426348925/05aeef6d-6432-4320-a521-025803848f49?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20231122%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20231122T193742Z&X-Amz-Expires=300&X-Amz-Signature=61e77906f86879af1ed4c6612b3c21322f0dd9e8cac3790684c42479bf35e9c0&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=426348925&response-content-disposition=attachment%3B%20filename%3Dbees-wasps.h5&response-content-type=application%2Foctet-stream [following]
--2023-11-22 20:37:42--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/426348925/05aeef6d-6432-4320-a521-02580384

## Question 1

Now convert this model from Keras to TF-Lite format. What is the size in MB of TF-Lite model?

In [10]:
import os
import tensorflow as tf
from tensorflow import keras

In [6]:
# Load the TensorFlow model
model_path = "bees-wasps.h5"
model = keras.models.load_model(model_path)

# Convert the TF Model to TF-Lite Model
converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()

# Saving the TF-Lite model
tflite_model_path = "bees-wasps_model.tflite"
with open(tflite_model_path, "wb") as f_out:
    f_out.write(tflite_model)

INFO:tensorflow:Assets written to: /tmp/tmptf2dxz2l/assets


INFO:tensorflow:Assets written to: /tmp/tmptf2dxz2l/assets
2023-11-22 20:41:53.614890: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:378] Ignored output_format.
2023-11-22 20:41:53.614907: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:381] Ignored drop_control_dependency.
2023-11-22 20:41:53.615035: I tensorflow/cc/saved_model/reader.cc:83] Reading SavedModel from: /tmp/tmptf2dxz2l
2023-11-22 20:41:53.615670: I tensorflow/cc/saved_model/reader.cc:51] Reading meta graph with tags { serve }
2023-11-22 20:41:53.615679: I tensorflow/cc/saved_model/reader.cc:146] Reading SavedModel debug info (if present) from: /tmp/tmptf2dxz2l
2023-11-22 20:41:53.617470: I tensorflow/cc/saved_model/loader.cc:233] Restoring SavedModel bundle.
2023-11-22 20:41:53.683947: I tensorflow/cc/saved_model/loader.cc:217] Running initialization op on SavedModel bundle at path: /tmp/tmptf2dxz2l
2023-11-22 20:41:53.691722: I tensorflow/cc/saved_model/loader.cc:316] SavedModel

In [17]:
print(f"sizeof({tflite_model_path}) = {os.stat(tflite_model_path).st_size/1000**2:.0f} MB")

sizeof(bees-wasps_model.tflite) = 45 MB


**Solution:**
* 21 Mb
* **`43 Mb`** (~45 MB)
* 80 Mb
* 164 Mb


## Question 2

To be able to use this model, we need to know the index of the input and 
the index of the output. 

What's the output index for this model?

In [18]:
import tensorflow.lite as tflite

In [19]:
interpreter = tflite.Interpreter(model_path=tflite_model_path)
interpreter.allocate_tensors()

INFO: Created TensorFlow Lite XNNPACK delegate for CPU.


In [20]:
input_index = interpreter.get_input_details()[0]["index"]
output_index = interpreter.get_output_details()[0]["index"]
print("input-index: ", input_index)
print("output-index: ", output_index)

input-index:  0
output-index:  13


**Solution:**
* 3
* 7
* **`13`**
* 24

## Preparing the image


You'll need some code for downloading and resizing images. You can use 
this code:


In [21]:
from io import BytesIO
from urllib import request

from PIL import Image

def download_image(url):
    '''Gets image from url and converts to PIL.Image'''
    with request.urlopen(url) as resp:
        buffer = resp.read()
    stream = BytesIO(buffer)
    img = Image.open(stream)
    return img

def prepare_image(img, target_size):
    if img.mode != 'RGB':
        img = img.convert('RGB')
    img = img.resize(target_size, Image.NEAREST)
    return img

Let's download and resize this image: 

In [50]:
url = "https://habrastorage.org/webt/rt/d9/dh/rtd9dhsmhwrdezeldzoqgijdg8a.jpeg"
input_size = 150  # target_size of image to be used in xception-model (from previous homework)

# Get the image
img = download_image(url)
img = prepare_image(img, target_size=(input_size, input_size))

## Question 3


Now we need to turn the image into numpy array and pre-process it. 

> Tip: Check the previous homework. What was the pre-processing 
> we did there?

After the pre-processing, what's the value in the first pixel, the R channel?

In [51]:
import numpy as np

In [52]:
# Preprocessing ([0, 255] uint8 -> [0, 1] float)
def preprocess_input(x):
    x /= 255.0
    return x

In [57]:
# PIL.Image to np.ndarray
x = np.array(img, dtype=np.float32)[None, ...]
X = preprocess_input(x)

print(f"First Red-pixel: {X[0, 0, 0, 0]:.7f}")

First Red-pixel: 0.9450980


**Solution**:
* 0.3450980
* 0.5450980
* 0.7450980
* **`0.9450980`**

## Question 4

Now let's apply this model to this image. What's the output of the model?

In [58]:
interpreter.set_tensor(input_index, X)
interpreter.invoke()
preds = interpreter.get_tensor(output_index)

In [64]:
print(f"prediction: {preds.squeeze():.3f}")

prediction: 0.659


**Solution:**
* 0.258
* 0.458
* **`0.658`**
* 0.858

## Prepare the lambda code

Now you need to copy all the code into a separate python file. You will 
need to use this file for the next two questions.

The script can be found here: [lambda_function.py](lambda_function.py).

In [65]:
# Testing from the notebook
from lambda_function import lambda_handler

event = {"url": "https://habrastorage.org/webt/rt/d9/dh/rtd9dhsmhwrdezeldzoqgijdg8a.jpeg"}
result = lambda_handler(event, None)
print(result)

[0.6589840650558472]


INFO: Created TensorFlow Lite XNNPACK delegate for CPU.


## Question 5

Download the base image `agrigorev/zoomcamp-bees-wasps:v2`. You can easily make it by using [docker pull](https://docs.docker.com/engine/reference/commandline/pull/) command.

In [66]:
!docker pull agrigorev/zoomcamp-bees-wasps:v2

v2: Pulling from agrigorev/zoomcamp-bees-wasps

68a79b8a: Already exists 
124cce46: Pulling fs layer 
8b038848: Pulling fs layer 
7e7c1be9: Pulling fs layer 
8c0b7487: Pulling fs layer 
0580071d: Pulling fs layer 
Digest: sha256:823f8536a45968f40ee3daf8a2da030b914912a382a4611610b3b84d36d2924c[4A
Status: Downloaded newer image for agrigorev/zoomcamp-bees-wasps:v2
docker.io/agrigorev/zoomcamp-bees-wasps:v2


In [70]:
# Size of base-image
!docker image ls | grep agrigorev | awk '{print $NF}'

662MB


**Solution:**


* 162 Mb
* 362 Mb
* **`662 Mb`**
* 962 Mb

## Question 6

Now let's extend this docker image, install all the required libraries
and add the code for lambda.

You don't need to include the model in the image. It's already included. 
The name of the file with the model is `bees-wasps-v2.tflite` and it's 
in the current workdir in the image (see the Dockerfile above for the 
reference).

Now run the container locally.

- Dockerfile used: [Dockerfile](Dockerfile)
- Test-Script: [test.py](test.py)

Building the docker-container:
```sh
docker build -t agrigorev/zoomcamp-bees-wasps:v2 .
```

Running the docker-container:
```sh
docker run -it --rm -p 8080:8080 agrigorev/zoomcamp-bees-wasps:v2
```

In [72]:
# Calling the docker container from the Notebook
import requests

url = 'http://localhost:8080/2015-03-31/functions/function/invocations'
data = {'url': 'https://habrastorage.org/webt/rt/d9/dh/rtd9dhsmhwrdezeldzoqgijdg8a.jpeg'}

result = requests.post(url, json=data).json()
print(f"{result[0]:.4f}")

0.4453


**Solution:**
* 0.2453
* **`0.4453`**
* 0.6453
* 0.8453

## Publishing it to AWS


Now you can deploy your model to AWS!

* Publish your image to ECR
* Create a lambda function in AWS, use the ECR image
* Give it more RAM and increase the timeout 
* Test it
* Expose the lambda function using API Gateway